src/main/kotlin/cacophony/controlflow/FunctionHandler.kt

```diff
@@ -40,8 +41,82 @@ class FunctionHandlerImpl(
+
+        val nodes: MutableList<CFGNode> = mutableListOf()
+
+        if (respectStackAlignment) {
+            // we push two copies of RSP to the stack and either leave them both there,
+            // or remove one of them via RSP assignment
+            val oldRSP = Register.VirtualRegister()
+            nodes.add(CFGNode.Assignment(oldRSP, CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP))))
+            for (i in 0..2) {
+                nodes.add(CFGNode.Push(CFGNode.VariableUse(oldRSP)))
+            }
+
+            // in an ideal world we would do something like "and rsp, ~15" or similar; for now this will do
+            // at the very least split the computation of (RSP + stackArguments.size % 2 * 8) % 16
+            // into two cases depending on the parity of stackArguments.size
+            nodes.add(
+                CFGNode.Assignment(
+                    Register.FixedRegister(X64Register.RSP),
+                    CFGNode.Addition(
+                        CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP)),
+                        CFGNode.Modulo(
+                            CFGNode.Addition(
+                                CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP)),
+                                CFGNode.Constant(stackArguments.size % 2 * 8),
+                            ),
+                            CFGNode.Constant(16),
+                        ),
+                    ),
+                ),
+            )
+        }
+
+        // in what order should we evaluate arguments? gcc uses reversed order
+        for ((argument, register) in registerArguments) {
+            nodes.add(CFGNode.Assignment(Register.FixedRegister(register), argument))
+        }
+        for ((argument, register) in stackArguments) {
+            nodes.add(CFGNode.Assignment(register, argument))
+        }
+
+        // is this indirection necessary?
+        for ((_, register) in stackArguments.reversed()) {
+            nodes.add(CFGNode.Push(CFGNode.VariableUse(register)))
+        }
+
+        nodes.add(CFGNode.Call(function))
+
+        if (stackArguments.isNotEmpty()) {
+            nodes.add(
+                CFGNode.Assignment(
+                    Register.FixedRegister(X64Register.RSP),
+                    CFGNode.Addition(
+                        CFGNode.VariableUse(Register.FixedRegister(X64Register.RSP)),
+                        CFGNode.Constant(8 * stackArguments.size),
+                    ),
+                ),
+            )
+        }
+
+        if (respectStackAlignment) {
+            // we could remove the operations from previous `if (stackArguments.isNotEmpty())` block
+            // via MemoryAccess, but for now the semantics are a bit unclear + it would introduce
+            // a few ifs, which we do not need at this point
+            nodes.add(CFGNode.Pop(Register.FixedRegister(X64Register.RSP)))
```

Comment:
I'm not sure if I understand, but you pushed RSP 2 times, but here you pop it only once.

Reply:
I pop it directly to RSP, which implicitly pops the second RSP entry if it is there.